### ISSUE TITLE
Deadlock when serialising reference to queue

### ISSUE THREAD

**Issue by jonashaag** [dask/distributed#3752]
In the code below, when using `x = Queue()` and submitting `dask.delayed(g)` (without `()`), both the client and scheduler seem to deadlock. This also applies to non-local clusters.

```py
if __name__ == '__main__':
    import dask.distributed

    client = dask.distributed.Client()

    # OK:
    #x = dask.distributed.Variable()
    #x = dask.distributed.Semaphore()
    #x = dask.distributed.Lock()

    # Deadlock:
    x = dask.distributed.Queue()

    def f():
        return 42

    def g():
        x
        return 42

    print(client.gather(client.compute(dask.delayed(f)())))
    # => 42
    print(client.gather(client.compute(dask.delayed(f))))
    # => <function f at 0x7f7ee5eb5730>
    print(client.gather(client.compute(dask.delayed(g)())))
    # => 42
    print(client.gather(client.compute(dask.delayed(g))))
    # => infinite hang in gather
```

**Comment by jrbourbeau** [comment]
Thanks for raising this issue and providing a nice minimal example @jonashaag! I'm able to reproduce the hanging behavior with the current `master` branch. It looks like the last `Future`, the one created from `client.compute(dask.delayed(g))`, does finish but the worker(s) are stuck for some reason (I'm getting `distributed.worker - DEBUG - Heartbeat skipped: channel busy` logging messages when running your example).

Is this something you're interested in looking into more? (no obligation though) If not, hopefully I or someone else can investigate more in the near future

**Comment by jonashaag** [comment]
No, not at the moment since we apparently can’t use Queue, Variable, etc. in our cluster anyways, probably due to networking issues.

**Comment by nils-braun** [comment]
I tried to look into this issue and I see a deadlock - however not at the workers but at the client.
The workers are still reachable and happy in my case.

I did some step-through debugging and I think this is what happens:
* when `client.compute(dask.delayed(g))` is called, the function `g` including the global variable (the queue) it needs are pickled and are sent to the worker. 
* however, you are not requesting the result of `g` (because you are calling it without the `()`), but only `g`. `g` itself will never be called. 
* The worker will unpickle the function `g` including the queue, will not do anything and store the unpickled version (I am not sure what exactly happens, but it is also not relevant).
* when the `gather` is called, a `sync` is called to get the result. This will pickle the function and the queue again on the worker and sent it to the client...
* ... where it gets unpickled again.

So far, so expected. However, unpickling a queue calls its initializer, which leads to a call to `sync` (https://github.com/dask/distributed/blob/master/distributed/queues.py#L186). This happens inside the call to `gather`, which is itself a call to `sync` (which will block forever as the inner sync will never be processed).

Unfortunately, I am not sure how to fix this (yet) because I do not have enough experience with event loops - maybe @jrbourbeau has some idea?
